binary_search_tree/binary_search_tree.py

```diff
@@ -0,0 +1,196 @@
+                    parent_node.right = None
+                elif node.right is None:
+                    # if node only has left child
+                    parent_node.right = node.left
+                elif node.left is None:
+                    # if node only has right child
+                    parent_node.right = node.right
+                else:
+                    # if node have both chidren
+                    if node.right.left is None:
+                        node.right.left = node.left
+                        parent_node.right = node.right
+                    else:
+                        to_replace = node.right
+                        parent_to_replace = node
+                        while to_replace.left is not None:
+                            parent_to_replace = to_replace
+                            to_replace = to_replace.left
+                        parent_node.right.data = to_replace.data
+                        parent_to_replace.left = to_replace.right
+                return
+
+    def search_node(self, data):
+        """
+        search element in binary tree
+        Time Complexity: O(Logn)average case and O(n) in worst case
+        :param data: search node of value data in bst
+        :return : parent of the node whose value is data or raise exception
+        """
+        parent_node = None
+        if self.root is None:
+            raise LookupError('Cannot delete anything, BST is empty')
+        elif self.root.data == data:
+            return parent_node
+        else:
+            iterator = self.root
+            while iterator is not None:
+                if iterator.data == data:
+                    return parent_node
+                elif iterator.data > data:
+                    parent_node = iterator
+                    iterator = iterator.left
+                else:
+                    parent_node = iterator
+                    iterator = iterator.right
+            raise LookupError('Node does not exist')
+
+    def print_preorder(self, node):
+        """
+        Print all element of binary tree in preorder using recurtion
+        """
+        if node is None:
+            return
+        elif node.left is None and node.right is None:
+            sys.stdout.write(str(node.data) + ' ')
+        else:
+            self.print_preorder(node.left)
+            sys.stdout.write(str(node.data) + ' ')
+            self.print_preorder(node.right)
+            if self.root == node:
+                sys.stdout.write('\n')
+            return
+
+
+def main():
+    inp = bst()
+    inp.insert_node(4)
+    inp.insert_node(8)
+    inp.insert_node(6)
+    inp.insert_node(5)
+    inp.insert_node(44)
+    inp.insert_node(69)
+    inp.print_preorder(inp.root)
+    inp.delete_node(69)
+    inp.delete_node(5)
+    inp.delete_node(4)  # delete root
+    inp.print_preorder(inp.root)
+    print(inp.root.data)  # check value of root
+    inp.delete_node(2)  # just to throw exception
```

Comment:
> Whats the problem with `raise`?

nothing with `raise`, I asked to put this line in `try-except` block due to which the exception will be caught and problem can have exit code 0. I hope I'm clear and correct for you to change this.